safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,182 @@
+
+// nonceSize is the size of the nonces in bytes.
+const nonceSize = 8
+
+func (pd *PolicyDirective) generateNonce() {
+	b := make([]byte, nonceSize)
+	_, err := pd.readRand(b)
+	if err != nil {
+		// TODO: handle this better, what should happen here?
+		panic(err)
+	}
+	pd.lastNonce = base64.StdEncoding.EncodeToString(b)
+}
+
+// Policy defines a CSP policy, containing many directives.
+type Policy struct {
+	Directives []*PolicyDirective
+}
+
+// NewPolicy creates a new strict, nonce-based CSP.
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+//
+// TODO: maybe reportURI should be safehttp.URL?
+func NewPolicy(reportURI string) *Policy {
+	return &Policy{
+		Directives: []*PolicyDirective{
+			NewPolicyDirective(DirectiveObjectSrc, []string{ValueNone}, false),
+			NewPolicyDirective(DirectiveScriptSrc, []string{
+				ValueUnsafeInline,
+				ValueUnsafeEval,
+				ValueStrictDynamic,
+				ValueHTTPS,
+				ValueHTTP,
+			}, true),
+			NewPolicyDirective(DirectiveBaseURI, []string{ValueNone}, false),
+			NewPolicyDirective(DirectiveReportURI, []string{reportURI}, false),
+		},
+	}
+}
+
+// Serialize serializes this policy for use in a Content-Security-Policy header
+// or in a Content-Security-Policy-Report-Only header. The nonces generated for
+// each directive is also returned.
+func (p Policy) Serialize() (csp string, nonces map[Directive]string) {
+	nonces = make(map[Directive]string)
+	b := strings.Builder{}
+
+	for i, d := range p.Directives {
+		if i != 0 {
+			b.WriteString("; ")
+		}
+		b.WriteString(string(d.dir))
+
+		if d.nonces {
+			d.generateNonce()
+			b.WriteString(" 'nonce-")
+			b.WriteString(d.lastNonce)
+			b.WriteString("'")
+			nonces[d.dir] = d.lastNonce
+		}
+
+		for _, v := range d.vals {
+			b.WriteString(" ")
+			b.WriteString(v)
+		}
+	}
+	return b.String(), nonces
+}
+
+// Interceptor intercepts requests and applies CSP policies.
+type Interceptor struct {
+	// EnforcementPolicy will be applied as the Content-Security-Policy header.
+	EnforcementPolicy *Policy
+
+	// ReportOnlyPolicy will be applied as the
+	// Content-Security-Policy-Report-Only header.
+	ReportOnlyPolicy *Policy
+}
+
+// Default creates a new CSP interceptor with a strict nonce-based policy in
+// enforcement mode.
+func Default(reportURI string) Interceptor {
+	return Interceptor{EnforcementPolicy: NewPolicy(reportURI)}
+}
+
+type ctxKey string
+
+// Before claims and sets the Content-Security-Policy header and the
+// Content-Security-Policy-Report-Only header.
+func (it Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	h := w.Header()
+	setCSP, err := h.Claim("Content-Security-Policy")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+	if it.EnforcementPolicy != nil {
+		s, nonces := it.EnforcementPolicy.Serialize()
+		setCSP([]string{s})
+		if len(nonces) != 0 {
+			r.SetContext(context.WithValue(r.Context(), ctxKey("enforce"), nonces))
+		}
```

Comment:
Why do you need to set anything in the context?

I assume this is for some future changes? In this case, I'd rather not add it yet. You can leave a `TODO` instead. 

Especially since as it is right now, the implementation could use some improvements, which you would likely notice once you use it.

Reply:
> which you would likely notice once you use it

Should I create a dummy server that uses all current plugins, so I can discover these potential issues? Or leave that for later?

Reply:
I'll do that now